### ISSUE TITLE
Identical SHAP values when TreeExplainer is applied to RandomForestClassifier

### ISSUE THREAD

**Issue by IvanUkhov** [shap/shap#102]
First of all, thank you for the great work!

As the title suggests, I’m trying to explain `sklearn.ensemble.RandomForestClassifier` using `shap.TreeExplainer`. If it’s a binary classification problem, `shap_values` returns a list of two matrices, one for each class. The problem is that the matrices are identical except for the last column, which corresponds to the bias. They both seem to try to explain the same class.

Here is an example:

```python
import sklearn
import shap

from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split

X, y = shap.datasets.iris()
y[y == 2] = 1

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=0)

model = RandomForestClassifier(n_estimators=100, max_depth=None,
                               min_samples_split=2, random_state=0)
model.fit(X_train, y_train)

explainer = shap.TreeExplainer(model)
shap_values = explainer.shap_values(X_test)

shap.initjs()
```

```python
y_test
```

```text
array([1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1,
       0, 0, 1, 0, 0, 1, 1, 0])
```

```python
shap.force_plot(shap_values[0], X_test)
```

<img width="999" alt="0" src="https://user-images.githubusercontent.com/57440/40575443-6accb6a0-60e5-11e8-9468-241812e1b739.png">

```python
shap.force_plot(shap_values[1], X_test)
```

<img width="1007" alt="1" src="https://user-images.githubusercontent.com/57440/40575446-711c4c50-60e5-11e8-937a-512be689a3ad.png">

The same applies to more than two classes.

If you switch to `LogisticRegression` and `KernelExplainer`, you’ll indeed get different SHAP values for different classes. In the case of two classes, they’ll be negations of each other (sum up to zero).

Is it a bug, or am I misusing the library? Thanks!

**Comment by slundberg** [comment]
When I run your code and then `shap_values[0][0,0], shap_values[1][0,0]` I get `(-0.024567362467574919, 0.024567362467574912)` which seems to be what we both expect. Do you get something different for that?

I also get the negative of what you get for `shap.force_plot(shap_values[1], X_test)`. What version of shap and sklearn are you using?

**Comment by IvanUkhov** [comment]
I’ve just repeated the experiment and got `(-0.02456736246757492, -0.02456736246757492)`. The software that I have installed is as follows:

* Python 3.5.2
* numpy 1.14.3 (also tried with 1.14.1)
* pandas 0.23.0 (also tried with 0.22.0)
* scikit-learn 0.19.1
* shap 0.16.1

**Comment by slundberg** [comment]
Strange. I have deadline in the next few days, but I'll let you know what I
find when I can dig into it.

**Comment by IvanUkhov** [comment]
I perfectly understand. Good luck with the deadline!

If you’d like to reproduce the result, start a Docker container running [`tensorflow/tensorflow:latest-py3`](https://hub.docker.com/r/tensorflow/tensorflow/) and then install `shap`.

**Comment by slundberg** [comment]
So it turns out there was a bug in the tree_shap.h file that didn't offset correctly. This caused multi-class problems to always output the SHAP values for the first class for all the other classes. Why it worked for me at first, and then didn't when I found the bug I am not sure. Something must have been updated that caused this. To prevent it from happening again I've added a unit test for it.

Thanks so much for pointing this out!
